ocean_prefilter/ocean_prefilter.go

```diff
@@ -0,0 +1,289 @@
+		rc.detector, err = vision.FromDependencies(deps, prefilterConfig.DetectorName)
+		if err != nil {
+			return errors.Wrapf(err, "unable to get camera %v for ocean prefilter", prefilterConfig.DetectorName)
+		}
+	}
+	rc.chosenLabels = prefilterConfig.ChosenLabels // if you configred an optional detector, this determines the labels and confidences to use
+
+	// now start the background thread
+	pf.activeBackgroundWorkers.Add(1)
+	viamutils.ManagedGo(func() {
+		// if you get an error while running just keep trying forever
+		for {
+			runErr := run(pf.cancelContext, rc, pf.triggerFlag)
+			if runErr != nil {
+				pf.logger.Error(runErr)
+				continue // keep trying to run, forever
+			}
+			return
+		}
+	}, func() {
+		pf.activeBackgroundWorkers.Done()
+	})
+	return nil
+}
+
+// run sets up a camera stream and then takes new pictures and processes them for anomalies
+// at the desired frequency.
+func run(ctx context.Context, rc runConfig, trigger *int32) error {
+	if rc.cam == nil {
+		return errors.Errorf("underlying camera %q is nil, cannot start background stream", rc.camName)
+	}
+	stream, err := rc.cam.Stream(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer stream.Close(ctx)
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+			start := time.Now()
+			img, release, err := stream.Next(ctx)
+			if err != nil {
+				atomic.StoreInt32(trigger, triggerFalse)
+				release()
+				return err
+			}
+			isTriggered, err := theFilter(img) // bogus stand in function for now
+			if isTriggered {
+				atomic.StoreInt32(trigger, triggerTrue)
+			} else {
+				atomic.StoreInt32(trigger, triggerFalse)
+			}
+			release()
+
+			done := time.Now()
+			took := done.Sub(start)
+			waitFor := time.Duration((1/rc.frequency)*float64(time.Second)) - took // only poll according to set freq
+			if waitFor > time.Microsecond {
+				time.Sleep(waitFor)
+			}
+		}
+	}
+}
+
+// theFilter is a stupid fillter function just for testing purposes
+func theFilter(img image.Image) (bool, error) {
+	// if the minute is even, return true
+	currentTime := time.Now()
+	minute := currentTime.Minute()
+	isEven := minute%2 == 0
+	return isEven, nil
+}
+
+func (pf *prefilter) DetectionsFromCamera(
+	ctx context.Context,
+	cameraName string,
+	extra map[string]interface{},
+) ([]objdet.Detection, error) {
+	return nil, errUnimplemented
+}
+
+func (pf *prefilter) Detections(ctx context.Context, img image.Image, extra map[string]interface{}) ([]objdet.Detection, error) {
+	return nil, errUnimplemented
+}
+
+func (pf *prefilter) ClassificationsFromCamera(
+	ctx context.Context,
+	cameraName string,
+	n int,
+	extra map[string]interface{},
+) (classification.Classifications, error) {
+	if cameraName != pf.camName {
+		return nil, errors.Errorf("Camera name given to method, %v is not the same as configured camera %v", cameraName, pf.camName)
+	}
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-pf.cancelContext.Done():
```

Comment:
If the run goroutine is closed, then there are no more images being processed - it means that the latest atomic.Bool will remain the same and quickly become scale. Let's say the goroutine exited and the trigger bool was in state TRUE. It would remain true until the goroutine restarted. I would instead want an error essentially saying that the background stream has closed and not return a stale classification

Reply:
Makes sense; thanks!